[release/9.0-staging] Fix few RandomAccess.Write edge case bugs

adamsitnik opened the pull request:
Backport of https://github.com/dotnet/runtime/pull/108380 and https://github.com/dotnet/runtime/pull/109340 and https://github.com/dotnet/runtime/pull/109826 to release/9.0-staging

## Customer Impact

- [x] Customer reported
- [x] Found internally

Customers using `RandomAccess.Write` overload that accepts multiple buffers have observed `IOException` being thrown in two scenarios:
- On Windows, when the file offset (stored as `Int32`) overflows and the method tries to perform a write with negative file offset. Reported by the user in https://github.com/dotnet/runtime/issues/108322.
- On Unix, when the number of buffers is larger than `IOV_MAX` (usually 1024). Reported by the user in https://github.com/dotnet/runtime/issues/108383.

When I was fixing  https://github.com/dotnet/runtime/issues/108383 I've realized that the logic for handling incomplete writes has a bug.
For incomplete multi-buffer writes, the API was not reporting any exceptions. It returns void, so it should write everything or throw an exception.

## Regression

- [ ] Yes
- [x] No

We had all 3 bugs since the API was introduced (.NET 6).

## Testing

New unit tests were added. They were failing before the fix were applied. They are passing now.

## Risk

Low, the fix is simple and covered with tests.

NicoAvanzDev committed:
RandomAccess.Write throws System.IO.IOException Fix (#108380)

adamsitnik committed:
Handle IOV_MAX limit in RandomAccess, add missing test (#109340)

* add test for Int32 overflow for WriteGather in RandomAccess

* add failing test fore more than IOV_MAX buffers

* fix both the native and managed parts

---------

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>
Co-authored-by: Stephen Toub <stoub@microsoft.com>

adamsitnik commented:
/azp list

adamsitnik commented:
/azp run runtime-libraries-coreclr outerloop

ericstj commented:
@adamsitnik did you want to port the same test fixes to this PR that you mentioned in https://github.com/dotnet/runtime/pull/109648?

adamsitnik commented:
> @adamsitnik did you want to port the same test fixes to this PR that you mentioned in https://github.com/dotnet/runtime/pull/109648?

@ericstj yes, and to be exact the fixes are #109826

ericstj commented:
Ok, please add back servicing-consider when it's ready.

adamsitnik committed:
More WriteGather fixes (#109826)

* don't run these tests in parallel, as each test cases uses more than 4 GB ram and disk!

* fix the test: handle incomplete reads that should happen when we hit the max buffer limit

* incomplete write fix:

- pin the buffers only once
- when re-trying, do that only for the actual reminder

* Use native memory to get OOM a soon as we run out of memory (hoping to avoid the process getting killed on Linux when OOM happens)

* For macOS preadv and pwritev can fail with EINVAL when the total length of all vectors overflows a 32-bit integer.

* add an assert that is going to warn us if vector.Count is ever more than Int32.MaxValue

---------

Co-authored-by: Michał Petryka <35800402+MichalPetryka@users.noreply.github.com>

adamsitnik commented:
/azp run runtime-libraries-coreclr outerloop

carlossanlop commented:
Friendly reminder that code complete is on April 14th for the May Release. If you'd like to get this change included in that release, please get a Tactics approval and merge this PR before that date.

stephentoub commented:
@adamsitnik, @jeffhandley, should this be closed?

adamsitnik committed:
Merge branch 'release/9.0-staging' into backportWriteGather9

stephentoub commented on src/native/libs/System.Native/pal_io.c:
```diff
@@ -1883,6 +1883,53 @@ int32_t SystemNative_PWrite(intptr_t fd, void* buffer, int32_t bufferSize, int64
+#if defined(IOV_MAX)
+    const int IovMax = IOV_MAX;
+#else
+    // In theory all the platforms that we support define IOV_MAX,
```
Should we just fail to build if it's not defined?

adamsitnik replied:
> Should we just fail to build if it's not defined?

We could end up breaking some configuration that is not maintained/built by us, which is something that I would prefer to avoid if possible.

tarekgh commented:
@adamsitnik are planning to get this to the current servicing release? If so, please try to merge this soon. 

Just reminder Code-complete is 14-Jul 4pm PT cut-off

adamsitnik commented:
/ba-g All test failures are known (WASM) and reported as such by BA
